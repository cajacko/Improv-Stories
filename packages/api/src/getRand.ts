function defaultGetRand() {
  return Math.random();
}

let getRand: () => number = defaultGetRand;

export function __TESTS__setGetRand(callback: typeof getRand) {
  getRand = callback;
}

export function __TESTS__reset() {
  getRand = defaultGetRand;
}

export default () => getRand();
