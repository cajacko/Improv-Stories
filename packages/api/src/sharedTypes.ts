export interface UserDetails {
  name: string | null;
}

export interface User extends UserDetails {
  dateModified: string;
  dateAdded: string;
  id: string;
  version: number;
}

export interface Session {
  id: string;
  user: User;
  dateStarted: string;
  dateWillFinish: string;
  finalEntry: string;
  entries: string[];
  dateModified: string;
  version: number;
}

export interface Story {
  id: string;
  connectedUsers: User[];
  activeUsers: User[];
  lastSession: null | Session;
  activeSession: null | Session;
  dateCreated: string;
  dateModified: string;
  version: number;
}

export interface Message<T, P = undefined> {
  id: string;
  type: T;
  payload: P;
  createdAt: string;
}

export type ClientMessage =
  | Message<"ADD_USER_TO_STORY", { storyId: string; isActive: boolean }>
  | Message<"REMOVE_USER_FROM_STORY", { storyId: string }>
  | Message<"SET_USER_DETAILS", { userDetails: UserDetails }>
  | Message<"ADD_ACTIVE_USER_TO_STORY", { storyId: string }>
  | Message<"REMOVE_ACTIVE_USER_FROM_STORY", { storyId: string }>
  | Message<"ADD_STORY_ENTRY", { storyId: string; entry: string }>;

export type ServerMessage = Message<"STORY_CHANGED", Story>;
