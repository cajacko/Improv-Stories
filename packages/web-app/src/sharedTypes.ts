export interface UserDetails {
  name: string | null;
}

export interface User extends UserDetails {
  dateModified: string;
  dateAdded: string;
  id: string;
  version: number;
}

export interface BaseSession {
  id: string;
  dateStarted: string;
  dateWillFinish: string;
  finalEntry: string;
  entries: string[];
  dateModified: string;
  version: number;
  finishedEarly?: boolean;
}

export interface ServerSession extends BaseSession {
  user: User;
  dateFinished: null | string;
}

export interface DatabaseSession extends BaseSession {
  userId: string;
  dateFinished: string;
}

export interface DatabaseStoryProps {
  secondsPerRound?: number;
  canUsersEndRoundEarly?: boolean;
  storyId: string;
  storyPropsDateCreated: string;
  storyPropsDateModified: string;
  storyPropsVersion: number;
}

export interface Story {
  id: string;
  connectedUsers: User[];
  activeUsers: User[];
  lastSession: null | ServerSession;
  activeSession: null | ServerSession;
  dateCreated: string;
  dateModified: string;
  version: number;
}

export interface Message<T, P = undefined> {
  id: string;
  type: T;
  payload: P;
  createdAt: string;
}

export type ClientMessage =
  | Message<
      "LIVE_STORY_ADD_USER_TO_STORY",
      { storyId: string; isActive: boolean }
    >
  | Message<"LIVE_STORY_REMOVE_USER_FROM_STORY", { storyId: string }>
  | Message<"SET_USER_DETAILS", { userDetails: UserDetails }>
  | Message<"LIVE_STORY_ADD_ACTIVE_USER_TO_STORY", { storyId: string }>
  | Message<"LIVE_STORY_REMOVE_ACTIVE_USER_FROM_STORY", { storyId: string }>
  | Message<"LIVE_STORY_SET_SESSION_TEXT", { storyId: string; text: string }>
  | Message<
      "LIVE_STORY_SET_SESSION_DONE",
      { storyId: string; sessionId: string }
    >;
// | Message<"STANDARD_STORY_REQUEST_TAKE_TURN", { storyId: string }>;

export type ServerMessage =
  | Message<"LIVE_STORY_STORY_CHANGED", Story>
  | Message<"LIVE_STORY_SESSION_CHANGED", ServerSession>;
