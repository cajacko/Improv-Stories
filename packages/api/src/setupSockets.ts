import * as socket from "socket.io";
import broadCastStoriesChanged from "./broadCastStoriesChanged";
import { addUser, removeUser, getGetDate } from "./store";
import { ClientMessage } from "./sharedTypes";
import handleClientMessage from "./handleClientMessage";

function setupSockets(io: socket.Server) {
  function onClientMessage(userId: string) {
    return (message: ClientMessage) => {
      handleClientMessage(userId, message);
    };
  }

  function onSocketConnect(sock: socket.Socket, userId: string) {
    const date = getGetDate()();

    const changedStoryIds = addUser({
      id: userId,
      name: null,
      dateAdded: date,
      dateModified: date,
      connectedStories: {},
      version: 0,
      socket: sock,
    });

    broadCastStoriesChanged(changedStoryIds);

    sock.on("message", onClientMessage(userId));
  }

  function onSocketDisconnect(userId: string) {
    const changedStoryIds = removeUser(userId);

    broadCastStoriesChanged(changedStoryIds);
  }

  function getUserIdFromSocket(sock: socket.Socket) {
    if (!sock) return null;
    if (!sock.request) return null;
    if (!sock.request._query) return null;
    if (!sock.request._query.user_id) return null;

    return sock.request._query.user_id;
  }

  io.on("connection", (sock) => {
    // The only place we grab the user id
    const userId = getUserIdFromSocket(sock);

    if (!userId) {
      sock.disconnect(true);
      return;
    }

    onSocketConnect(sock, userId);

    sock.on("disconnect", () => {
      onSocketDisconnect(userId);
    });
  });
}

export default setupSockets;
