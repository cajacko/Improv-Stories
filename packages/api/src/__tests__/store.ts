import { v4 as uuid } from "uuid";
import { ClientMessage, ServerMessage } from "../sharedTypes";
import {
  __TESTS__setGetDatabase as setDatabase,
  __TESTS__reset as resetDatabase,
} from "../getDatabase";
import {
  __TESTS__reset as resetStore,
  __TESTS__getStore as getStore,
  __TESTS__setGetDate as setGetDate,
  __TESTS__setGetId as setGetId,
} from "../store";
import { __TESTS__setGetRand as setGetRand } from "../getRand";

import setupSockets from "../setupSockets";

jest.useFakeTimers();

function connectUserAndGetFuncs(userId: string) {
  let sendClientMessage: (
    message: Omit<ClientMessage, "id" | "createdAt">
  ) => Promise<any>;
  let disconnect: () => {};

  const sendMock = jest.fn<unknown, [ServerMessage]>();

  const sock: any = {
    // NOTE: Keep this as we use to use this for the id, but now don't and want to ensure it doesn't
    // ruin anything
    id: uuid(),
    request: {
      _query: {
        user_id: userId,
      },
    },
    on: (type: any, callback: any) => {
      if (type === "disconnect") {
        disconnect = callback;
      } else {
        sendClientMessage = callback;
      }
    },
    send: sendMock,
  };

  let connectUser: any;

  const io: any = {
    on: (type: any, callback: any) => {
      connectUser = callback;
    },
  };

  setupSockets(io);

  connectUser(sock);

  return {
    // @ts-ignore
    sendClientMessage,
    // @ts-ignore
    disconnect,
    sendMock,
  };
}

type Funcs = ReturnType<typeof connectUserAndGetFuncs>;

const resetDateValue = new Date(2020, 1, 1, 0, 0, 0, 0);

function getUserSendCall(
  user: Funcs,
  callNumber: number,
  doNotAsset = false
): ServerMessage {
  const call = user.sendMock.mock.calls[callNumber - 1];

  if (!doNotAsset) {
    expect(!!call).toBe(true);
  }

  return call[0];
}

function getUserSendCallFromEnd(
  user: Funcs,
  callFromEndNumber: number = 0,
  doNotAsset = false
): ServerMessage {
  const callNumber = user.sendMock.mock.calls.length - callFromEndNumber;

  return getUserSendCall(user, callNumber, doNotAsset);
}

function getLastUserSendCall(user: Funcs, doNotAsset = false): ServerMessage {
  return getUserSendCallFromEnd(user, 0, doNotAsset);
}

function expectUserSendCount(user: Funcs, count: number) {
  expect(user.sendMock.mock.calls.length).toBe(count);
}

function expectSameUserSentMessagesFromEnd(
  users: Funcs[],
  callFromEndNumber: number = 0,
  doNotAsset = false
) {
  expect(users.length > 1).toBe(true);

  const firstCall = getUserSendCallFromEnd(
    users[0],
    callFromEndNumber,
    doNotAsset
  );

  users.forEach((user) => {
    expect(firstCall.payload).toEqual(
      getUserSendCallFromEnd(user, callFromEndNumber, doNotAsset).payload
    );

    expect(firstCall.type).toEqual(
      getUserSendCallFromEnd(user, callFromEndNumber, doNotAsset).type
    );
  });
}

function expectSameUserLastSentMessages(users: Funcs[], doNotAsset = false) {
  return expectSameUserSentMessagesFromEnd(users, 0, doNotAsset);
}

beforeEach(() => {
  resetStore();
  resetDatabase();

  setDatabase(() => ({
    ref: () => ({
      push: () => Promise.resolve(),
      once: (type: string, callback: any) => callback({ val: () => undefined }),
    }),
  }));

  const date = new Date(resetDateValue);

  setGetDate(() => {
    date.setMinutes(date.getMinutes() + 1);
    return date.toISOString();
  });

  let id = 0;

  setGetId(() => {
    id += 1;
    return `${id}`;
  });

  let rand = 1;

  setGetRand(() => {
    rand += 1;
    if (rand >= 10) rand = 1;

    return rand / 10;
  });
});

afterEach(() => {
  jest.clearAllTimers();
});

describe("user1 connects", () => {
  const user1Id = "user1";
  let user1: Funcs;

  beforeEach(() => {
    user1 = connectUserAndGetFuncs(user1Id);
  });

  afterEach(() => {
    if (user1) {
      user1.disconnect();
    }
  });

  describe("user1 added to story1", () => {
    const story1Id = "story1";

    beforeEach(() => {
      user1.sendClientMessage({
        type: "LIVE_STORY_ADD_USER_TO_STORY",
        payload: {
          storyId: story1Id,
        },
      });
    });

    it("Broadcasts the correct changes to user1", () => {
      expect(getLastUserSendCall(user1)).toMatchSnapshot();
    });

    it("the user1 broadcast count is 1", () => {
      expectUserSendCount(user1, 1);
    });

    describe("user2 connects", () => {
      const user2Id = "user2";
      let user2: Funcs;

      beforeEach(() => {
        user2 = connectUserAndGetFuncs(user2Id);
      });

      afterEach(() => {
        if (user2) {
          user2.disconnect();
        }
      });

      describe("user2 added to story1", () => {
        beforeEach(() => {
          user2.sendClientMessage({
            type: "LIVE_STORY_ADD_USER_TO_STORY",
            payload: {
              storyId: story1Id,
            },
          });
        });

        it("Broadcasts the correct changes to user2", () => {
          expect(getLastUserSendCall(user2)).toMatchSnapshot();
        });

        it("the user2 broadcast count is 1", () => {
          expectUserSendCount(user2, 1);
        });

        it("Broadcasts the same changes to user1", () => {
          expectSameUserLastSentMessages([user1, user2]);
        });

        it("the user1 broadcast count is 2", () => {
          expectUserSendCount(user1, 2);
        });

        describe("user1 disconnects", () => {
          beforeEach(() => {
            user1.disconnect();
          });

          it("Broadcasts the correct changes to user2", () => {
            expect(getLastUserSendCall(user2)).toMatchSnapshot();
          });

          it("the user2 broadcast count is 2", () => {
            expectUserSendCount(user2, 2);
          });

          describe("user1 connects", () => {
            beforeEach(() => {
              user1 = connectUserAndGetFuncs(user1Id);
            });

            afterEach(() => {
              if (user1) {
                user1.disconnect();
              }
            });

            describe("user1 added to story1", () => {
              const story1Id = "story1";

              beforeEach(() => {
                user1.sendClientMessage({
                  type: "LIVE_STORY_ADD_USER_TO_STORY",
                  payload: {
                    storyId: story1Id,
                  },
                });
              });

              it("Broadcasts the correct changes to user1", () => {
                expect(getLastUserSendCall(user1)).toMatchSnapshot();
              });

              it("the user1 broadcast count is 1", () => {
                expectUserSendCount(user1, 1);
              });

              it("Broadcasts the same changes to user2", () => {
                expectSameUserLastSentMessages([user1, user2]);
              });

              it("the user2 broadcast count is 3", () => {
                expectUserSendCount(user2, 3);
              });
            });
          });
        });

        describe("user1 becomes active on story1", () => {
          beforeEach(() => {
            user1.sendClientMessage({
              type: "LIVE_STORY_ADD_ACTIVE_USER_TO_STORY",
              payload: {
                storyId: story1Id,
              },
            });
          });

          it("Broadcasts the correct changes to user2", () => {
            expect(getLastUserSendCall(user2)).toMatchSnapshot();
          });

          it("the user2 broadcast count is 2", () => {
            expectUserSendCount(user2, 2);
          });

          it("Broadcasts the same changes to user1", () => {
            expectSameUserLastSentMessages([user1, user2]);
          });

          it("the user1 broadcast count is 3", () => {
            expectUserSendCount(user1, 3);
          });

          describe("user2 becomes active on story1", () => {
            beforeEach(() => {
              user2.sendClientMessage({
                type: "LIVE_STORY_ADD_ACTIVE_USER_TO_STORY",
                payload: {
                  storyId: story1Id,
                },
              });
            });

            it("Broadcasts the correct changes to user2", () => {
              expect(getUserSendCallFromEnd(user2, 1)).toMatchSnapshot();
              expect(getLastUserSendCall(user2)).toMatchSnapshot();
            });

            it("the user2 broadcast count is 4", () => {
              expectUserSendCount(user2, 4);
            });

            it("Broadcasts the same changes to user1", () => {
              expectSameUserLastSentMessages([user1, user2]);
              expectSameUserSentMessagesFromEnd([user1, user2], 1);
            });

            it("the user1 broadcast count is 5", () => {
              expectUserSendCount(user1, 5);
            });

            it("user1 is the active user", () => {
              const message = getLastUserSendCall(user2);

              if (message.type !== "LIVE_STORY_STORY_CHANGED") {
                throw new Error("Expected message type to be STORY_CHANGED");
              }

              if (!message.payload.activeSession) {
                throw new Error(
                  "Expected the message payload to have activeSession"
                );
              }

              expect(message.payload.activeSession.user.id).toBe("user1");
            });

            describe("user1 sets text", () => {
              beforeEach(() => {
                user1.sendClientMessage({
                  type: "LIVE_STORY_SET_SESSION_TEXT",
                  payload: {
                    text: "a",
                    storyId: story1Id,
                  },
                });
              });

              it("Broadcasts the correct changes to user2", () => {
                expect(getLastUserSendCall(user2)).toMatchSnapshot();
              });

              it("the user2 broadcast count is 5", () => {
                expectUserSendCount(user2, 5);
              });

              it("Broadcasts the same changes to user1", () => {
                expectSameUserLastSentMessages([user1, user2]);
                expectSameUserSentMessagesFromEnd([user1, user2], 1);
              });

              it("the user1 broadcast count is 6", () => {
                expectUserSendCount(user1, 6);
              });

              // TODO: Test session elapsing, saving to db and switching users here
              // TODO: Test setting session as done

              describe("user1 ends the session early", () => {
                beforeEach(() => {
                  return user1.sendClientMessage({
                    type: "LIVE_STORY_SET_SESSION_DONE",
                    payload: {
                      sessionId: "8",
                      storyId: story1Id,
                    },
                  });
                });

                it("Broadcasts the correct changes to user2", () => {
                  expect(getLastUserSendCall(user2)).toMatchSnapshot();
                });

                it("the user2 broadcast count is 6", () => {
                  expectUserSendCount(user2, 6);
                });

                it("Broadcasts the same changes to user1", () => {
                  expectSameUserLastSentMessages([user1, user2]);
                  expectSameUserSentMessagesFromEnd([user1, user2], 1);
                });

                it("the user1 broadcast count is 7", () => {
                  expectUserSendCount(user1, 7);
                });

                // TODO: Check db saved, user id change is correct
              });
            });

            describe("user1 disconnects", () => {
              beforeEach(() => {
                user1.disconnect();
              });

              it("Broadcasts the correct changes to user2", () => {
                expect(getLastUserSendCall(user2)).toMatchSnapshot();
              });

              it("the user2 broadcast count is 5", () => {
                expectUserSendCount(user2, 5);
              });

              it("the active session is the same as before", () => {
                const message1 = getLastUserSendCall(user1);
                const message2 = getLastUserSendCall(user2);

                if (message1.type !== "LIVE_STORY_STORY_CHANGED") {
                  throw new Error("Expected message type to be STORY_CHANGED");
                }

                if (!message1.payload.activeSession) {
                  throw new Error(
                    "Expected the message payload to have activeSession"
                  );
                }

                if (message2.type !== "LIVE_STORY_STORY_CHANGED") {
                  throw new Error("Expected message type to be STORY_CHANGED");
                }

                if (!message2.payload.activeSession) {
                  throw new Error(
                    "Expected the message payload to have activeSession"
                  );
                }

                expect(message1.payload.activeSession.id).toEqual(
                  message2.payload.activeSession.id
                );
              });
            });

            describe("user2 disconnects", () => {
              beforeEach(() => {
                user2.disconnect();
              });

              it("Broadcasts the correct changes to user1", () => {
                expect(getLastUserSendCall(user1)).toMatchSnapshot();
              });

              it("the user1 broadcast count is 6", () => {
                expectUserSendCount(user1, 6);
              });

              it("the active session is the same as before", () => {
                const message1 = getLastUserSendCall(user1);
                const message2 = getLastUserSendCall(user2);

                if (message1.type !== "LIVE_STORY_STORY_CHANGED") {
                  throw new Error("Expected message type to be STORY_CHANGED");
                }

                if (!message1.payload.activeSession) {
                  throw new Error(
                    "Expected the message payload to have activeSession"
                  );
                }

                if (message2.type !== "LIVE_STORY_STORY_CHANGED") {
                  throw new Error("Expected message type to be STORY_CHANGED");
                }

                if (!message2.payload.activeSession) {
                  throw new Error(
                    "Expected the message payload to have activeSession"
                  );
                }

                expect(message1.payload.activeSession.id).toEqual(
                  message2.payload.activeSession.id
                );
              });
            });
          });
        });
      });
    });
  });
});
