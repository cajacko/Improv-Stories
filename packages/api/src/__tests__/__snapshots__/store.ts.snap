// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user1 connects user1 added to story1 Broadcasts the correct changes to user1 1`] = `
Object {
  "createdAt": "2020-02-01T00:04:00.000Z",
  "id": "1",
  "payload": Object {
    "activeSession": null,
    "activeUsers": Array [],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:03:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 2,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 Broadcasts the correct changes to user2 1`] = `
Object {
  "createdAt": "2020-02-01T00:08:00.000Z",
  "id": "3",
  "payload": Object {
    "activeSession": null,
    "activeUsers": Array [],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:06:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 3,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 user1 becomes active on story1 Broadcasts the correct changes to user2 1`] = `
Object {
  "createdAt": "2020-02-01T00:11:00.000Z",
  "id": "5",
  "payload": Object {
    "activeSession": null,
    "activeUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
    ],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:09:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 4,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 user1 becomes active on story1 user2 becomes active on story1 Broadcasts the correct changes to user2 1`] = `
Object {
  "createdAt": "2020-02-01T00:14:00.000Z",
  "id": "7",
  "payload": Object {
    "activeSession": null,
    "activeUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:12:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 5,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 user1 becomes active on story1 user2 becomes active on story1 Broadcasts the correct changes to user2 2`] = `
Object {
  "createdAt": "2020-02-01T00:19:00.000Z",
  "id": "10",
  "payload": Object {
    "activeSession": Object {
      "dateModified": "2020-02-01T00:15:00.000Z",
      "dateStarted": "2020-02-01T00:15:00.000Z",
      "dateWillFinish": "2020-02-01T00:15:21.000Z",
      "entries": Array [],
      "finalEntry": "",
      "id": "8",
      "user": Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      "version": 0,
    },
    "activeUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:17:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 7,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 user1 becomes active on story1 user2 becomes active on story1 user1 disconnects Broadcasts the correct changes to user2 1`] = `
Object {
  "createdAt": "2020-02-01T00:24:00.000Z",
  "id": "11",
  "payload": Object {
    "activeSession": Object {
      "dateModified": "2020-02-01T00:15:00.000Z",
      "dateStarted": "2020-02-01T00:15:00.000Z",
      "dateWillFinish": "2020-02-01T00:15:21.000Z",
      "entries": Array [],
      "finalEntry": "",
      "id": "8",
      "user": Object {
        "dateAdded": "2020-02-01T00:22:00.000Z",
        "dateModified": "2020-02-01T00:23:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      "version": 0,
    },
    "activeUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:21:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 9,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 user1 becomes active on story1 user2 becomes active on story1 user2 disconnects Broadcasts the correct changes to user1 1`] = `
Object {
  "createdAt": "2020-02-01T00:22:00.000Z",
  "id": "11",
  "payload": Object {
    "activeSession": Object {
      "dateModified": "2020-02-01T00:15:00.000Z",
      "dateStarted": "2020-02-01T00:15:00.000Z",
      "dateWillFinish": "2020-02-01T00:15:21.000Z",
      "entries": Array [],
      "finalEntry": "",
      "id": "8",
      "user": Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
      "version": 0,
    },
    "activeUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
    ],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:01:00.000Z",
        "dateModified": "2020-02-01T00:01:00.000Z",
        "id": "user1",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:21:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 9,
  },
  "type": "STORY_CHANGED",
}
`;

exports[`user1 connects user1 added to story1 user2 connects user2 added to story1 user1 disconnects Broadcasts the correct changes to user2 1`] = `
Object {
  "createdAt": "2020-02-01T00:11:00.000Z",
  "id": "4",
  "payload": Object {
    "activeSession": null,
    "activeUsers": Array [],
    "connectedUsers": Array [
      Object {
        "dateAdded": "2020-02-01T00:05:00.000Z",
        "dateModified": "2020-02-01T00:05:00.000Z",
        "id": "user2",
        "name": null,
        "version": 0,
      },
    ],
    "dateCreated": "2020-02-01T00:02:00.000Z",
    "dateModified": "2020-02-01T00:10:00.000Z",
    "id": "story1",
    "lastSession": null,
    "version": 5,
  },
  "type": "STORY_CHANGED",
}
`;
